### PR TITLE
Disallow deletion of composable template if in use by data stream

### LINKED ISSUE #57004: Prevent removing a template when it is used by a data stream
This validation is required in order to avoid that a data stream gets rolled over and then fail, because there isn't a matching template with a data stream template snippet.

### PR DESCRIPTION
**Pull request by dakrone**
Currently a data stream requires a component template in order to be created. This is so that the
stream can be perpetuated when the index is automatically or manually rolled over. This commit
ensures that the composable index template used by the stream is not removed if the data stream
still exists.

Resolves #57004

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java [hunk 1]
```diff
@@ -617,6 +629,18 @@ static ClusterState innerRemoveIndexTemplateV2(ClusterState currentState, String
         return ClusterState.builder(currentState).metadata(metadata).build();
     }
 
+    static Set<String> dataStreamsUsingTemplate(final ClusterState state, final String templateName) {
+        final Set<String> dataStreams = state.metadata().dataStreams().keySet();
```
Comment (danhermann) [inline server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java:633]: pico-optimization: `dataStreams` not needed until after `templateName` is confirmed to exist

**Review by danhermann — APPROVED** [review]
LGTM, eggplant references aside. ;)

**Review by martijnvg — APPROVED** [review]
LGTM 👍

**Comment by dakrone** [comment]
@elasticmachine update branch

**Pull request merged**